kernel: fix cross compilation

lopsided98 opened the pull request:
This PR contains rebased versions of patches from #30882 to fix Linux kernel cross compilation.

Specifically, this includes:
5c6473c5ff4879711d28243f4be299380ead10fb - kernel: Enable cross compiling
d1ccc5d0b48321498a57bfb372fb782701446ad7 - kernel: Use build kmod
5e917600fc61f996c297aeeeb8a231b75460dfd4 - kernel fixes
e915e4a7d0cd798070b969517310acaaf378bd39 - kernel: Fix cross-compilation

5e917600fc61f996c297aeeeb8a231b75460dfd4 has been squashed into 5c6473c5ff4879711d28243f4be299380ead10fb because the changes are closely related.

There are a few things that I'm not sure about, or didn't seem right to me.

In 5c6473c5ff4879711d28243f4be299380ead10fb, `pkgs/os-specific/linux/kernel/perf.nix` refers to a [patch that no longer exists in master](https://github.com/NixOS/nixpkgs/pull/30882/commits/5c6473c5ff4879711d28243f4be299380ead10fb#diff-ec4e01e2f58d1c9120aad93a3dde92e4R16) and did not exist in the tree that #30882 is based on. It also seems to revert the [fix that was used to replace the patch](https://github.com/NixOS/nixpkgs/pull/30882/commits/5c6473c5ff4879711d28243f4be299380ead10fb#diff-ec4e01e2f58d1c9120aad93a3dde92e4L45). This seems like it could be due to a bad rebase at some point, but I wanted to make sure there wasn't another reason.

I put a few other concerns as comments on the diff.

I have tested cross building 4.14 kernels for `armv7l` and `aarch64`, as well as a native `x86_64` kernel.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bgamari committed:
kernel: Enable cross compiling

bgamari committed:
kernel: Use build kmod

bgamari committed:
kernel: Fix cross-compilation

lopsided98 commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -233,13 +225,13 @@ let
-      } // extraMeta;
-    };
+      };
+    } // extraMeta;
```
This doesn't make sense to me. Why shouldn't `extraMeta` be merged with the existing `meta` attribute?

lopsided98 commented on pkgs/os-specific/linux/kernel/perf.nix:
```diff
@@ -23,8 +23,8 @@ stdenv.mkDerivation {
-      flex bison libiberty libaudit makeWrapper pkgconfig ];
-  buildInputs = [ elfutils python perl newt slang libunwind binutils zlib ] ++
+      flex bison libiberty libaudit makeWrapper pkgconfig python perl ];
+  buildInputs = [ elfutils newt slang libunwind binutils zlib ] ++
```
I'm not sure if these dependencies are all correct.

teto commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -58,37 +58,29 @@ let
 
   configfile = stdenv.mkDerivation {
-    inherit ignoreConfigErrors;
+    #inherit ignoreConfigErrors;
```
leftover ?

lopsided98 commented:
cc @bgamari @Ericson2314

lopsided98 commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -58,37 +58,29 @@ let
+    arch = hostPlatform.platform.kernelArch;
+
+    # TODO(@Ericson2314): No null next hash break
+    ignoreConfigErrors = if stdenv.hostPlatform == stdenv.buildPlatform then null else true;
```
What does "No null next hash break" mean?

Also, I just discovered this breaks builds with config errors, because `ignoreConfigErrors` does not get defined when not cross compiling and the scripts assume it is defined.

lopsided98 committed:
kernel: don't force ignoreConfigErrors when cross compiling, only set the
default

teto commented:
Quite excited to see some improvements on the kernel as it's the most annoying software to work with on nixos. Getting rid of crossplatform-specific code makes patching easir (I have some patches to handle the kernel config in a simpler way but I could not upstream them because they didn't support cross-platform).
Could your rebase this so that I can test it please ?

teto commented:
I rebased locally (you need to pay attention to https://github.com/NixOS/nixpkgs/pull/33057/commits/b1ca8517ee54b89c24c77af3b9408066cee57d74 ) but then I get
```
@[nix-shell:~/nixops]$ nixops deploy -d 828db7f5-f9a1-11e7-a652-309c233b770e --show-trace
error: while evaluating anonymous function at /home/teto/nixpkgs/lib/options.nix:116:41, called from /home/teto/nixpkgs/lib/attrsets.nix:199:52:
while evaluating ‘scrubOptionValue’ at /home/teto/nixpkgs/lib/options.nix:112:22, called from /home/teto/nixpkgs/lib/options.nix:116:44:
while evaluating ‘isDerivation’ at /home/teto/nixpkgs/lib/attrsets.nix:295:18, called from /home/teto/nixpkgs/lib/options.nix:113:8:
while evaluating the attribute ‘baseImage’ at /home/teto/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/teto/nixpkgs/lib/modules.nix:75:45, called from /home/teto/nixpkgs/lib/attrsets.nix:199:52:
while evaluating the attribute ‘value’ at /home/teto/nixpkgs/lib/modules.nix:312:9:
while evaluating the option `deployment.libvirtd.baseImage':
while evaluating the attribute ‘mergedValue’ at /home/teto/nixpkgs/lib/modules.nix:339:5:
while evaluating anonymous function at /home/teto/nixpkgs/lib/modules.nix:339:32, called from /home/teto/nixpkgs/lib/modules.nix:339:19:
while evaluating ‘check’ at /home/teto/nixpkgs/lib/types.nix:318:15, called from /home/teto/nixpkgs/lib/modules.nix:340:10:
while evaluating ‘check’ at /home/teto/nixpkgs/lib/types.nix:230:15, called from /home/teto/nixpkgs/lib/types.nix:318:31:
while evaluating the attribute ‘args’ of the derivation ‘libvirtd-ssh-image’ at /home/teto/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘text’ of the derivation ‘vm-run’ at /home/teto/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘outPath’ at /home/teto/nixpkgs/lib/customisation.nix:155:7:
while evaluating the attribute ‘configurePhase’ of the derivation ‘linux-4.9.76’ at /home/teto/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘arch’ of the derivation ‘linux-config-4.9.76’ at /home/teto/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
attribute ‘platform’ missing, at /home/teto/nixpkgs/pkgs/os-specific/linux/kernel/generic.nix:75:12
```
which is strange since in repl, hostPlatform.platform does exist.

teto commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -122,22 +114,17 @@ let
+  #configWithPlatform = kernelPlatform: import ./common-config.nix
+  #  { inherit stdenv version kernelPlatform extraConfig;
+  #    features = passthru.features; # Ensure we know of all extra patches, etc.
+  #  };
```
leftover

teto commented:
I pushed https://github.com/teto/nixpkgs/tree/kernel_cross which seems ok for host builds. would be nice to check cross platform.

bgamari commented:
Thanks for doing this!

lopsided98 committed:
kernel: remove leftover remnants of old cross compiling system

lopsided98 committed:
perf: inherit makeFlags from kernel

lopsided98 committed:
kernel: revert seemingly incorrect change to extraMeta merging

lopsided98 commented:
I rebased and added some new commits to fix some cosmetic issues and remove some dead code left over from the old cross compiling system. Most of these changes were not hash breaking, but I did revert a change @bgamari made to how `extraMeta` was merged (80f6b8e, it didn't seem right, but let me know if it was that way for a reason), and I made perf inherit its `makeFlags` from the kernel. I was unable to test cross building perf because some of its dependencies don't build yet, but a native build worked.

Another thing I was wondering about is why do we need to ignore config errors when cross compiling? Config errors are already ignored on all non "pc" host platforms, so the only case I can think of where the new cross compiling condition would be used is when cross compiling on Darwin for x86_64 Linux.

Ericson2314 commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -65,31 +64,14 @@ let
-        kernelBaseConfig = cp.kernelBaseConfig;
-        kernelTarget = cp.kernelTarget;
-        autoModules = cp.kernelAutoModules;
+    nativeBuildInputs = [ buildPackages.stdenv.cc perl ];
```
Should be a depsBuildBuild

Ericson2314 commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -239,34 +231,25 @@ in
 
-  nativeBuildInputs = [ perl bc nettools openssl gmp libmpc mpfr ]
-      ++ optional (stdenv.platform.kernelTarget == "uImage") ubootTools
+  nativeBuildInputs = [ perl bc nettools openssl gmp libmpc mpfr buildPackages.stdenv.cc ]
```
buildPackages.stdenv.cc should again be a depsBuildBuild.

Ericson2314 commented on pkgs/os-specific/linux/kernel/perf.nix:
```diff
@@ -23,8 +23,8 @@ stdenv.mkDerivation {
   # perf refers both to newt and slang
   # binutils is required for libbfd.
```
There is a libbfd derivation now.

Ericson2314 reviewed (commented):
Thanks for doing this!

lopsided98 committed:
kernel: fix dependencies

teto commented:
Thanks for the patch.
sorry to hijack the thread but as the kernel configuration is system dependant (which makes tailoring the kernel much harder, buildLinux should accept the parameters directly and the system forcefully enable some but I digress), I have my kernel config defined in a custom platform 

```
  test-platform = super.platforms.pc64_simplekernel // {
    kernelAutoModules = false;
    extraConfig = super.platforms.pc64_simplekernel
      + kvmConfig
      + debugConfig
      ;
  };
```
My understanding is that I need to crosscompile for the same architecture just to override some kernel settings :/ ?! I've never done any crosscompilation, what could be the command to compile the kernel with the previous platform please (in a nix-shell for instance) ?

NB: the naming hostPlatform.platform is confusing since hostPlatform looks like a platform already, shouldn't it be renamed to hostSystem or accept directly the platform argument ?

Ericson2314 commented:
All the `*Platform`s have a `platform` field, for the record. I agree its an ugly grab-bag of program-specific settings though, and I hope to gradually get rid of it.

You don't need to cross compile, you can pass `localSystem` to nixpkgs which has the same structure as the `*Platform`s. I'd write you an example now, but I'm on my phone. The manual is now a bit better in this stuff though.

teto commented:
I managed to come up with
`nix-build -A neovim  --arg localSystem '(import <nixpkgs/lib>).systems.examples.aarch64-multiplatform' '<nixpkgs>' --show-trace` thanks to the manual (I wish there were more examples though).
My system is not aarch64 though. Is it possible to override the platform component of the localSystem ? I believe it's defined in `lib/systems/default.nix` but as it's an impure thing I am not sure how to do this. My nix-info:
` system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 1.11.16,`

Ericson2314 commented:
@teto Yes, the syntax of `--arg` is `--arg <identifier> <nixexpr>` so you can put whatever you want in the second part.

teto commented:
Sorry I am a bit dense, I first tried this:
`nix-build -A neovim  --arg 'localSystem.platform' '(import <nixpkgs/lib>).systems.platforms.raspberrypi' '<nixpkgs>' --show-trace ` but it doesnt recompile anything. I believe the identifier might be wrong as I've read some issues where it mentioned specific rules.
In pseudo code I would like to do sthg like
`nix-build -A neovim  --arg 'localSystem' 'localSystem.platform // {  kernelArch = "arm"; }' '<nixpkgs>' -vvv `, i.e. not having to define localSystem from scratch.

NB: I would like to add your answer to the doc PR I prepare https://github.com/NixOS/nixpkgs/pull/29944

teto commented:
My example was bad, I was more interested in changing the platform like ` platform = lib.systems.platforms.pc64_simplekernel; }'` which works with your anser \o/
I am glad I asked since the answer is not so trivial. I'll add it to the manual as it is handy until kernel config becomes easier. Once again my apologies for having hijacked the thread.

Ericson2314 commented:
@teto no worries and yeah it indeed is cumbersome. So if we were to remove the parameters from platform as you suggested, I'd love, if possible, to make it so the scenario I pointed out is impossible. For example maybe `kernelArch` could just be computed on the fly from `hostPlatform.parsed.arch`, whereas other things would be callPackage parameters.

vcunat commented:
This caused me quite a headache.  Can some of you have a look at https://github.com/NixOS/nixpkgs/pull/34882 ?  I'm not good in the cross stuff...
